File: src/cpp/25_napisy_trudne/zad9.cpp
/*
Tytul: Najdluzszy wspolny podnapis.
Tresc: Dostajesz dwa napisy. Znajdz najdluzszy wspolny podnapis tych dwoch
napisow. Dane wejsciowe: Dwa napisy. Dane wyjsciowe: Napis. Przyklad: Dla
napisow "ijkabcdl" i "xxxxabcd" powinien zostac zwrocony napis "abcd".

*/
#include <cassert>
#include <string>
#include <vector>

std::string najdluzszyPodnapisV1(const std::string &slowoA,
                                 const std::string &slowoB) {

  auto m = slowoA.length();
  auto n = slowoB.length();

  std::vector<std::vector<int>> pom(m, vector<int>(n));

  for (auto i = 1; i <= m; i++) {
    for (auto j = 1; j <= n; j++) {
      if (slowoA[i - 1] == slowoB[j - 1])
        pom[i][j] = pom[i - 1][j - 1] + 1;
      else
        pom[i][j] = 0;
    }
  }

  std::string wynik = "";
  for (auto i = 1; i <= m; i++) {
    for (auto j = 1; j <= n; j++) {
      if (pom[i][j] > wynik.length())
        wynik = slowoA.substr((i - pom[i][j] + 1) - 1, pom[i][j]);
    }
  }
  return wynik;
}

// Testy Poprawnosci
void test1() {
  std::string slowoA = "abcdxyz";
  std::string slowoB = "xyzabcd";
  std::string wynik = "abcd";

  assert(najdluzszyPodnapisV1(slowoA, slowoB) == wynik);
}

void test2() {
  std::string slowoA = "hhaall";
  std::string slowoB = "hhaall";
  std::string wynik = "hhaall";

  assert(najdluzszyPodnapisV1(slowoA, slowoB) == wynik);
}

void test3() {
  std::string slowoA;
  std::string slowoB;
  std::string wynik;

  assert(najdluzszyPodnapisV1(slowoA, slowoB) == wynik);
}

int main() {

  test1();
  test2();
  test3();

  return 0;
}

File: src/python/25_napisy_trudne/zad9.py
"""
Tytul: Najdluzszy wspolny przedrostek.
Tresc: Dostajesz liste napisow. Znajdz najdluzszy wspolny przedrostek dla wszystkich napisow na liscie.
Dane wejsciowe: Lista napisow.
Dane wyjsciowe: Napis.
Przyklad:
Dla listy ["Remolada", "Remux", "Remmy"] powinien zostac zwrocony napis "Rem".

"""

# Wersja 1
def najdluzszy_podnapis_v1(slowo_a, slowo_b):

    m = len(slowo_a)
    n = len(slowo_b)

    pom = [[0 for i in range(n + 1)] for j in range(m + 1)]

    for i in range(1, m + 1):
        for j in range(1, n + 1):
            if slowo_a[i - 1] == slowo_b[j - 1]:
                pom[i][j] = pom[i - 1][j - 1] + 1
            else:
                pom[i][j] = 0

    wynik = ""
    for i in range(1, m + 1):
        for j in range(1, n + 1):
            if pom[i][j] > len(wynik):
                wynik = slowo_a[(i - pom[i][j] + 1) - 1 : pom[i][j]]

    return wynik


# Testy Poprawnosci
def test_1():
    slowo_a = "abcdxyz"
    slowo_b = "xyzabcd"
    wynik = "abcd"
    assert najdluzszy_podnapis_v1(slowo_a, slowo_b) == wynik


def test_2():
    slowo_a = "hhaall"
    slowo_b = "hhaall"
    wynik = "hhaall"
    assert najdluzszy_podnapis_v1(slowo_a, slowo_b) == wynik


def test_3():
    slowo_a = ""
    slowo_b = ""
    wynik = ""
    assert najdluzszy_podnapis_v1(slowo_a, slowo_b) == wynik


def main():
    test_1()
    test_2()
    test_3()


if __name__ == "__main__":
    main()

